### PR TITLE
More documentation for the drm code

### PR DESCRIPTION
**Pull request by Drakulix**

### CODE REVIEW

File: src/backend/drm/atomic/mod.rs [hunk 1]
```diff
@@ -2,11 +2,21 @@
 //! [`RawDevice`](RawDevice) and [`RawSurface`](RawSurface)
 //! implementations using the atomic mode-setting infrastructure.
 //!
+//! Atomic mode-setting (previously referred to a nuclear page-flip) is a new api of the direct rendering
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:5]:
```suggestion
//! Atomic mode-setting (previously referred to a nuclear page-flip) is a new api of the Direct Rendering
//! Manager subsystem of the linux kernel. Adaptations of this api can be found in BSD kernels.
```

File: src/backend/drm/atomic/mod.rs [hunk 2]
```diff
@@ -2,11 +2,21 @@
...
+//! manager api of the linux kernel. Adaptions of this api can be found in BSD kernels.
+//!
+//! This api is objectively better then the outdated legacy-api, but no supported by every driver.
+//! Initialization will fail, if the api is unsupported. The legacy-api is also wrapped by smithay
+//! and may be used instead in these cases. Currently there are no features in smithay, that are
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:10]:
```suggestion
//! and may be used instead in these cases. Currently there are no features in smithay that are
```

File: src/backend/drm/atomic/mod.rs [hunk 3]
```diff
@@ -86,9 +96,11 @@ impl<A: AsRawFd + 'static> Drop for Dev<A> {
             // Here we restore the card/tty's to it's previous state.
             // In case e.g. getty was running on the tty sets the correct framebuffer again,
             // so that getty will be visible.
+            // We do exit correctly, if this fails, but the user will be presented with
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:99]:
```suggestion
            // We do exit correctly if this fails, but the user will be presented with
```

File: src/backend/drm/atomic/mod.rs [hunk 4]
```diff
@@ -86,9 +96,11 @@ impl<A: AsRawFd + 'static> Drop for Dev<A> {
...
+            // a black screen, if no display handler takes control again.
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:100]:
```suggestion
            // a black screen if no display handler takes control again.
```

File: src/backend/drm/atomic/mod.rs [hunk 5]
```diff
@@ -213,6 +225,8 @@ impl<A: AsRawFd + 'static> AtomicDrmDevice<A> {
 
         let dev_id = fstat(fd.as_raw_fd()).map_err(Error::UnableToGetDeviceId)?.st_rdev;
 
+        // we wrap some of the internal state in another struct to share with
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:228]:
```suggestion
        // we wrap some of the internal states in another struct to share with
```

File: src/backend/drm/atomic/mod.rs [hunk 6]
```diff
@@ -223,13 +237,22 @@ impl<A: AsRawFd + 'static> AtomicDrmDevice<A> {
             logger: log.clone(),
         };
 
-        // we want to modeset, so we better be the master, if we run via a tty session
+        // we want to modeset, so we better be the master, if we run via a tty session.
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:240]:
```suggestion
        // we need to be the master to do modesetting if we run via a tty session.
```

File: src/backend/drm/atomic/mod.rs (hunk 2 above)
Comment (Drakulix) [reply src/backend/drm/atomic/mod.rs:11]: atomic-fixup

File: src/backend/drm/atomic/mod.rs [hunk 7]
```diff
@@ -2,11 +2,22 @@
 //! [`RawDevice`](RawDevice) and [`RawSurface`](RawSurface)
 //! implementations using the atomic mode-setting infrastructure.
 //!
+//! Atomic mode-setting (previously referred to a nuclear page-flip) is a new api of the Direct Rendering
+//! Manager subsystem of the linux kernel. Adaptations of this api can be found in BSD kernels.
+//! manager api of the linux kernel. Adaptions of this api can be found in BSD kernels.
```
Comment (m-hugo) [inline src/backend/drm/atomic/mod.rs:7]:
```suggestion
```

File: src/backend/drm/mod.rs [hunk 8]
```diff
@@ -1,38 +1,153 @@
...
 //!
-//! ---
-//!
-//! Initialization of devices happens through an open file descriptor
-//! of a drm device.
+//! # Abstractions
 //!
-//! ---
+//! This is a model of what the user typically perceives as a `Device` and a `Surface`.
+//! The meaning of these phrases is very overloaded in the general computer graphics context,
```
Comment (vberger) [inline src/backend/drm/mod.rs:8]:
```suggestion
//! The meaning of these words is very overloaded in the general computer graphics context,
```

File: src/backend/drm/mod.rs [hunk 9]
```diff
@@ -1,38 +1,153 @@
...
+//! ## Surface
+//!
+//! A surface is a part of a `Device` that may output a picture to a number of connectors. It pumps pictures of buffers to outputs.
 //!
 //! On surface creation a matching encoder for your `encoder`-`connector` is automatically selected,
 //! if it exists, which means you still need to check your configuration.
 //!
-//! At last a [`Mode`](drm::control::Mode) needs to be selected,
-//! supported by the `crtc` in question.
+//! A surface consists of one `crtc` that is rendered to be the user. This is fixed for the `Surface`s lifetime and cannot be changed.
```
Comment (vberger) [inline src/backend/drm/mod.rs:54]:
```suggestion
//! A surface consists of one `crtc` that is rendered to by the user. This is fixed for the `Surface`s lifetime and cannot be changed.
```

File: src/backend/drm/mod.rs [hunk 10]
```diff
@@ -1,38 +1,153 @@
...
+//! A commit/page_flip may be triggered by any other method the `Surface` has, but needs to be documented as such.
+//! The most low-level `Surface`s also implement `RawSurface`, which let you trigger a commit explicitly.
+//!
+//! ## RawDevice
+//!
+//! A low-level device that produces not only `Surface`s, but `RawSurface`s.
+//!
+//! ## RawSurface
+//!
+//! RawSurfaces provide two additional functions related to rendering. `commit` and `page_flip`.
```
Comment (vberger) [inline src/backend/drm/mod.rs:69]:
```suggestion
//! RawSurfaces provide two additional functions related to rendering: `commit` and `page_flip`.
```

File: src/backend/drm/mod.rs [hunk 11]
```diff
@@ -1,38 +1,153 @@
...
+//!
+//! To perform actual rendering operations you need to render your image to a framebuffer.
+//! Several types of framebuffers exists.
+//!
+//! The simplest one - the `Dumbbuffer` is a simple bitmap, usually very slow and not hardware-accelerated.
+//! It should be avoided at any cost, but works on all devices and can be used directly with a `RawSurface`.
+//! Rendering to it can be done manually by coping bitmaps around (cpu rendering).
+//!
+//! Accelerated buffers are usually hardware manufacture dependent. A problem, which is mostly solved by
+//! the General Buffer Manager (gbm-)API. `gbm::Buffer`s can only be rendered to by egl. Appropriate hardware-accelerated
```
Comment (vberger) [inline src/backend/drm/mod.rs:86]: > [...] can only be rendered to by egl

Is GBM incompatible with Vulkan, or is it just that Smithay does not have Vulkan support yet?

File: src/backend/drm/mod.rs [hunk 12]
```diff
@@ -1,38 +1,153 @@
...
+//! buffers are picked automatically. gbm even may fall back to `Dumbbuffers` on unsupported devices and
+//! egl-support is then provided by Mesa's software renderer `llvmpipe`.
+//!
+//! An alternative api to `gbm` is the so called `eglstream`-api, proposed (and implemented) by nvidia.
+//! It is currenly only supported by nvidias proprietary driver, which is also not accelerated by gbm, although other
+//! manifactures could in theory implement the standard, just like nvidia could implement gbm-support in their driver.
```
Comment (vberger) [inline src/backend/drm/mod.rs:92]:
```suggestion
//! manufactures could in theory implement the standard, just like nvidia could implement gbm-support in their driver.
```

File: src/backend/drm/mod.rs [hunk 13]
```diff
@@ -1,38 +1,153 @@
...
+//! - [`EglStreamDevice`](eglstream::EglStreamDevice) (and [`EglStreamSurface`](eglstream::EglStreamSurface)), which replace the direct rendering capabilities of any underlying `RawSurface` with functionality to create EGLStreams.
+//!
+//! On top of that the [`EglDevice`](egl::EglDevice) (and [`EglSurface`](egl::EglSurface)) replace the manual buffer management
+//! with an implementation of smithays [`GLGraphicsBackend`](::backend::graphics::gl::GLGraphicsBackend) to initialize the OpenGL-API for rendering.
+//!
+//! Additionally the [`FallbackDevice`](fallback::FallbackDevice) provides an easy wrapper to automatically pick an appropriate device.
+//! E.g. it can initialize an `AtomicDrmDevice` or fallback to a `LegacyDrmDevice` with the atomic api is not supported by the graphics card.
+//! It can also check for a loaded nvidia-driver for a given device and choose between an `EglStreamDevice` or a `GbmDevice`.
+//!
+//! Any layer of this chain can be implemented and therefor extended by the user. All building blocks are exposed to add addtional
```
Comment (vberger) [inline src/backend/drm/mod.rs:118]:
```suggestion
//! Any layer of this chain can be implemented and therefore extended by the user. All building blocks are exposed to add addtional
```

File: src/backend/drm/mod.rs [hunk 14]
```diff
@@ -1,38 +1,153 @@
...
+//! buffer management solutions or rendering apis out of tree. Additionally all of these implementations can be excluded from a build
+//! by disabling their corresponding feature.
+//!
+//! The most versatile type correctly provided by smithay, that should be able to initialize almost every
```
Comment (vberger) [inline src/backend/drm/mod.rs:122]:
```suggestion
//! The most versatile type currently provided by smithay, that should be able to initialize almost every
```
I think that's what you meant?

File: src/backend/drm/legacy/mod.rs [hunk 15]
```diff
@@ -2,11 +2,20 @@
 //! [`RawDevice`](RawDevice) and [`RawSurface`](RawSurface)
 //! implementations using the legacy mode-setting infrastructure.
 //!
+//! Legacy mode-setting referrs to the now outdated, but still supported direct manager api
```
Comment (vberger) [inline src/backend/drm/legacy/mod.rs:5]:
```suggestion
//! Legacy mode-setting refers to the now outdated, but still supported direct manager api
```

File: src/backend/drm/legacy/mod.rs [hunk 16]
```diff
@@ -2,11 +2,20 @@
...
+//! of the linux kernel. Adaptations of this api can be found in BSD kernels.
+//!
+//! The newer and objectively better api is known as atomic-modesetting (or nuclear-page-flip),
+//! however this api is not support by every driver, so this is provided for backwards compatibility.
```
Comment (vberger) [inline src/backend/drm/legacy/mod.rs:9]:
```suggestion
//! however this api is not supported by every driver, so this is provided for backwards compatibility.
```

File: src/backend/drm/legacy/session.rs [hunk 17]
```diff
@@ -72,6 +72,8 @@ impl<A: AsRawFd + 'static> LegacyDrmDeviceObserver<A> {
                 for surface in backends.borrow().values().filter_map(Weak::upgrade) {
                     // other ttys that use no cursor, might not clear it themselves.
                     // This makes sure our cursor won't stay visible.
+                    //
+                    // This usually happens with getty, and a cursor onto of a kernel console looks very weird.
```
Comment (vberger) [inline src/backend/drm/legacy/session.rs:76]:
```suggestion
                    // This usually happens with getty, and a cursor on top of a kernel console looks very weird.
```

File: src/backend/drm/atomic/mod.rs [hunk 18]
```diff
@@ -2,6 +2,15 @@
...
+//!
+//! This api is objectively better then the outdated legacy-api, but no supported by every driver.
```
Comment (vberger) [inline src/backend/drm/atomic/mod.rs:9]:
```suggestion
//! This api is objectively better than the outdated legacy-api, but not supported by every driver.
```

File: src/backend/drm/atomic/mod.rs [hunk 19]
```diff
@@ -225,13 +238,22 @@ impl<A: AsRawFd + 'static> AtomicDrmDevice<A> {
...
 
-        // enable the features we need
+        // Enable the features we need.
+        // We technically could use the atomic api without universal plane support.
+        // But the two are almost exclusively implemented together, as plane synchronization
+        // is one of the killer-features of the atomic api.
+        //
+        // We could bould more abstractions in smithay for devices with partial support,
+        // but for now we role with the oldest possible api (legacy) and the newest feature set
+        // we can untilize (atomic + universal planes), although we barely use planes yet.
```
Comment (vberger) [inline src/backend/drm/atomic/mod.rs:256]:
```suggestion
        // we can use (atomic + universal planes), although we barely use planes yet.
```

File: src/backend/drm/atomic/session.rs [hunk 20]
```diff
@@ -75,6 +75,8 @@ impl<A: AsRawFd + 'static> AtomicDrmDeviceObserver<A> {
             for surface in backends.borrow().values().filter_map(Weak::upgrade) {
                 // other ttys that use no cursor, might not clear it themselves.
                 // This makes sure our cursor won't stay visible.
+                //
+                // This usually happens with getty, and a cursor onto of a kernel console looks very weird.
```
Comment (vberger) [inline src/backend/drm/atomic/session.rs:79]:
```suggestion
                // This usually happens with getty, and a cursor on top of a kernel console looks very weird.
```

File: src/backend/drm/mod.rs (hunk 12 above)
Comment (m-hugo) [reply src/backend/drm/mod.rs:92]:
```suggestion
//! manufacturers could in theory implement the standard, just like nvidia could implement gbm-support in their driver.
```

File: src/backend/drm/mod.rs (hunk 11 above)
Comment (Drakulix) [reply src/backend/drm/mod.rs:86]: Last time I looked there was no Vulkan extension in Mesa to create vulkan objects from gbm ones. Such an extension (like the one existing for egl) would be necessary to use Vulkan on top of gbm.

As it stands there is no universally supported extension, so I assume most users of gbm will only see egl as a viable option.

Comment (vberger) [reply src/backend/drm/mod.rs:86]: Alright then.

File: src/backend/drm/mod.rs (hunk 14 above)
Comment (Drakulix) [reply src/backend/drm/mod.rs:122]: yep

File: src/backend/drm/mod.rs [hunk 21]
```diff
@@ -1,38 +1,153 @@
...
+//! Rendering to it can be done manually by coping bitmaps around (cpu rendering).
+//!
+//! Accelerated buffers are usually hardware manufacture dependent. A problem, which is mostly solved by
+//! the General Buffer Manager (gbm-)API. `gbm::Buffer`s can only be rendered to by egl. Appropriate hardware-accelerated
+//! buffers are picked automatically. gbm even may fall back to `Dumbbuffers` on unsupported devices and
+//! egl-support is then provided by Mesa's software renderer `llvmpipe`.
+//!
+//! An alternative api to `gbm` is the so called `eglstream`-api, proposed (and implemented) by nvidia.
+//! It is currenly only supported by nvidias proprietary driver, which is also not accelerated by gbm, although other
+//! manufactures could in theory implement the standard, just like nvidia could implement gbm-support in their driver.
```
Comment (m-hugo) [inline src/backend/drm/mod.rs:92]:
```suggestion
//! manufacturers could in theory implement the standard, just like nvidia could implement gbm-support in their driver.
```

**Pull request merged**